### PR TITLE
SCMOD-8630: Allow Elasticsearch protocol to be configurable

### PR DESCRIPTION
**Pull request by thilagasanthosh**
https://portal.digitalsafe.net/browse/SCMOD-8630
http://sou-jenkins2.swinfra.net:8080/job/CAFAudit/job/CAFAudit~audit-service~SCMOD-8630~CI/

### CODE REVIEW

File: caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditRestHighLevelClientFactory.java [hunk 1]
```diff
@@ -50,6 +50,11 @@ private ElasticAuditRestHighLevelClientFactory() {
      */
     public static RestHighLevelClient getHighLevelClient(final String hostAndPortValues)
         throws AuditConfigurationException {
+        String elasticProtocol = System.getProperty(ElasticAuditConstants.ConfigEnvVar.CAF_ELASTIC_PROTOCOL,
+                System.getenv(ElasticAuditConstants.ConfigEnvVar.CAF_ELASTIC_PROTOCOL));
+        if (elasticProtocol == null) {
+            elasticProtocol = ElasticAuditConstants.ConfigDefault.CAF_ELASTIC_PROTOCOL;
+        }
```
Comment (dermot-hardy) [inline caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditRestHighLevelClientFactory.java]: Shouldn't the configuration be read in the same place where all the other configuration is read?

File: caf-audit-monkey/README.md [hunk 2]
```diff
@@ -83,6 +83,11 @@ The following parameters may be set as required:
     <td>road.runner@acme.com, [Any String]</td>
     <td>Configurable field, available to the user. User who triggered the Audit Event</td>
   </tr>
+  <tr>
+     <td>CAF_ELASTIC_PROTOCOL</td>
+     <td>http, http</td>
```
Comment (dermot-hardy) [inline caf-audit-monkey/README.md]: Ahem

File: caf-audit/README.md [hunk 3]
```diff
@@ -140,7 +140,9 @@ A generated client-side library should be referenced in the normal way in the ap
 	</dependency>
 
 #### Direct to Elasticsearch Configuration  
-Configuration required to be supplied via environment variable:  
+Configuration required to be supplied via environment variable: 
+- `CAF_ELASTIC_PROTOCOL`: The protocol used to connect with elasticsearch server. e.g. http. Default value is http.
```
Comment (dermot-hardy) [inline caf-audit/README.md]: Mention https also

File: caf-audit-service-container/README.md [hunk 4]
```diff
@@ -6,6 +6,9 @@ This is a docker container for the [CAF Audit Web Service](https://github.com/CA
 
 #### Environment Variables
 
+##### CAF\_ELASTIC\_PROTOCOL
+The protocol used to connect with elasticsearch server. e.g. http. Default value is http.
```
Comment (dermot-hardy) [inline caf-audit-service-container/README.md]: Mention https also

File: docs/pages/en-us/Getting-Started.md [hunk 5]
```diff
@@ -258,6 +258,7 @@ This object represents a logical connection to the datastore (that is, Elasticse
 
 The `AuditConnection` object, for direct to Elasticseach, can be constructed by setting the `CAF_AUDIT_MODE` environment variable to `elasticsearch` and then using one of the static `createConnection()` methods in the `AuditConnectionFactory` class:  
 - `createConnection()` method has no parameters and requires the following environment variables to be set when creating a connection for direct to Elasticsearch:
+        - `CAF_ELASTIC_PROTOCOL`: The protocol used to connect with elasticsearch server. e.g. http. Defaults to http.
```
Comment (dermot-hardy) [inline docs/pages/en-us/Getting-Started.md]: - None of the others have a colon
- Mention https also

File: docs/pages/en-us/Client-API.md [hunk 6]
```diff
@@ -25,7 +25,8 @@ The order of instantiation and use of these objects for sending audit events is
 4. Use the [`AuditEventBuilder`](#AuditEventBuilder) object to construct and send audit event messages to the endpoint.
 
 #### Direct to Elasticsearch Configuration  
-Configuration required to be supplied via environment variable:  
+Configuration required to be supplied via environment variable:
+- `CAF_ELASTIC_PROTOCOL`: The protocol used to connect with elasticsearch server. e.g. http. Default value is http.
```
Comment (dermot-hardy) [inline docs/pages/en-us/Client-API.md]: Mention https also

File: caf-audit/README.md [hunk 7]
```diff
@@ -140,7 +140,9 @@ A generated client-side library should be referenced in the normal way in the ap
 	</dependency>
 
 #### Direct to Elasticsearch Configuration  
-Configuration required to be supplied via environment variable:  
+Configuration required to be supplied via environment variable:
```
Comment (dermot-hardy) [inline caf-audit/README.md]: Why is this first line changing?

File: deploy/src/main/resources/production-swarm/README.md [hunk 8]
```diff
@@ -58,4 +60,4 @@ To deploy the stack:
 * The CAF Audit Web Service containers will start up
 
 To tear down the stack:
-* Execute `docker stack rm auditWebServiceStack`
\ No newline at end of file
+* Execute `docker stack rm auditWebServiceStack`
```
Comment (dermot-hardy) [inline deploy/src/main/resources/production-swarm/README.md]: Why is this line changing?

File: docs/pages/en-us/Client-API.md [hunk 9]
```diff
@@ -25,7 +25,8 @@ The order of instantiation and use of these objects for sending audit events is
 4. Use the [`AuditEventBuilder`](#AuditEventBuilder) object to construct and send audit event messages to the endpoint.
 
 #### Direct to Elasticsearch Configuration  
-Configuration required to be supplied via environment variable:
```
Comment (dermot-hardy) [inline docs/pages/en-us/Client-API.md:28]: Why is this first line changing?

File: deploy/src/main/resources/production-marathon/environment.sh [hunk 10]
```diff
@@ -3,6 +3,9 @@
 # The port that the CAF Audit Web Service is configured to listen on.
 export CAF_AUDIT_SERVICE_PORT=
 
+# The protocol used to connect with elasticsearch server.
```
Comment (dermot-hardy) [inline deploy/src/main/resources/production-marathon/environment.sh]: The protocol used to connect **to the Elasticsearch server**.

File: deploy/src/main/resources/production-swarm/README.md [hunk 11]
```diff
@@ -19,6 +19,8 @@ The `environment.sh` file supports configurable property settings necessary for
 	###
 	# Elasticsearch
 	###
+	## The protocol used to connect with elasticsearch server.
```
Comment (dermot-hardy) [inline deploy/src/main/resources/production-swarm/README.md]: The protocol used to connect **to the Elasticsearch server**.

File: deploy/src/main/resources-filtered/docker-compose.yml [hunk 12]
```diff
@@ -107,6 +107,7 @@ services:
     ports:
       - "${CAF_AUDIT_SERVICE_PORT:-25080}:8080"
     environment:
+      CAF_ELASTIC_PROTOCOL: http
```
Comment (dermot-hardy) [inline deploy/src/main/resources-filtered/docker-compose.yml]: Move down to just above hosts.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: deploy/src/main/resources/production-marathon/environment.sh (hunk 10 above)
Comment (thilagasanthosh) [reply deploy/src/main/resources/production-marathon/environment.sh]: Updated description.

File: deploy/src/main/resources/production-swarm/README.md (hunk 11 above)
Comment (thilagasanthosh) [reply deploy/src/main/resources/production-swarm/README.md]: Updated description.

File: deploy/src/main/resources-filtered/docker-compose.yml (hunk 12 above)
Comment (thilagasanthosh) [reply deploy/src/main/resources-filtered/docker-compose.yml]: Elastic variables grouped together

File: docs/pages/en-us/Client-API.md (hunk 9 above)
Comment (thilagasanthosh) [reply docs/pages/en-us/Client-API.md:28]: removed change in the first line

File: deploy/src/main/resources/production-swarm/README.md (hunk 8 above)
Comment (thilagasanthosh) [reply deploy/src/main/resources/production-swarm/README.md]: Fixed

File: caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditRestHighLevelClientFactory.java (hunk 1 above)
Comment (thilagasanthosh) [reply caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditRestHighLevelClientFactory.java]: Initially, I added along with other environment variables section in ElasticAuditConnection. But that made changes to add an extra parameter to getHighLevelClient. Due to this extra parameter, there are quite a few test classes that need to be updated.  

For one of these variables, I don't want to touch many places in test classes. But now I moved it to ElasticAuditConnection class.

File: caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditConnection.java [hunk 13]
```diff
@@ -35,6 +35,11 @@
 
     public ElasticAuditConnection() throws AuditConfigurationException
     {
+            String elasticProtocol = System.getProperty(ElasticAuditConstants.ConfigEnvVar.CAF_ELASTIC_PROTOCOL,
+                    System.getenv(ElasticAuditConstants.ConfigEnvVar.CAF_ELASTIC_PROTOCOL));
+            if (elasticProtocol == null) {
+                elasticProtocol = ElasticAuditConstants.ConfigDefault.CAF_ELASTIC_PROTOCOL;
+            }
```
Comment (dermot-hardy) [inline caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditConnection.java]: Create a function for this.  Note that the fact that you can't mark `elasticProtocol` as `final` should have been giving you a strong hint that functional decomposition is required.

File: caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java [hunk 14]
```diff
@@ -51,12 +51,14 @@
...
     private static String ES_HOSTNAME_AND_PORT;
     private static int ES_PORT;
+    private static String CAF_ELASTIC_PROTOCOL;
 
     @BeforeClass
     public static void setup() throws Exception {
         // Test the Auditing library in direct mode
         System.setProperty("CAF_AUDIT_MODE", "elasticsearch");
-
+
```
Comment (dermot-hardy) [inline caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java]: Remove pointless whitespace added.

File: caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java [hunk 15]
```diff
@@ -321,4 +323,4 @@ private static void deleteIndex(RestHighLevelClient client, String indexId)
             }
         }
     }
-}
\ No newline at end of file
+}
```
Comment (dermot-hardy) [inline caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java]: Not part of this PR

File: caf-audit-service-container/src/test/java/com/hpe/caf/services/audit/api/WebserviceClientAuditIT.java [hunk 16]
```diff
@@ -81,7 +81,8 @@
     private static String WS_HOSTNAME;
     private static String WS_ENDPOINT;
     private static int WS_PORT;
-
+
```
Comment (dermot-hardy) [inline caf-audit-service-container/src/test/java/com/hpe/caf/services/audit/api/WebserviceClientAuditIT.java]: Remove pointless whitespace added.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
Why would you submit a PR for me to review when the build is failing?

File: caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditConnection.java (hunk 13 above)
Comment (thilagasanthosh) [reply caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditConnection.java]: The code to new function

File: caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java (hunk 14 above)
Comment (thilagasanthosh) [reply caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java]: Fixed

File: caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java (hunk 15 above)
Comment (thilagasanthosh) [reply caf-audit-binding-elasticsearch/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java]: Fixed

File: caf-audit-service-container/src/test/java/com/hpe/caf/services/audit/api/WebserviceClientAuditIT.java (hunk 16 above)
Comment (thilagasanthosh) [reply caf-audit-service-container/src/test/java/com/hpe/caf/services/audit/api/WebserviceClientAuditIT.java]: Fixed

File: caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditConnection.java [hunk 17]
```diff
@@ -104,6 +104,15 @@ public AuditChannel createChannel() throws IOException {
         return new ElasticAuditChannel(restHighLevelClient);
     }
 
+    private String getElasticProtocol(){
+        String elasticProtocol = System.getProperty(ElasticAuditConstants.ConfigEnvVar.CAF_ELASTIC_PROTOCOL,
```
Comment (dermot-hardy) [inline caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditConnection.java]: Re-write the function so that you don't have non-`final` variables in it.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Review by dermot-hardy — APPROVED** [review]
